### ISSUE TITLE
Client 12.6.0 uninstall leaves files behind and causes 12.7.2 install to fail when upgrading inside a chef-client run

### ISSUE THREAD

**Issue by Stromweld** [chef/chef#4623]
## Description

Trying to update chef-client 12.6.0 to 12.7.2(latest) fails when executed inside a chef-client run due to access denied errors on files during chef-client 12.6.0 uninstall phase.
## Chef Version

12.6.0 and upgrading to 12.7.2
## Platform Version

Windows 2012 r2 with chef-client 12.6.0 bootstrapped in.
## Replication Case

Trying to add feature to the omnibus_updater cookbook to add support for updating windows chef-clients. Using this repo https://github.com/hw-cookbooks/omnibus_updater.git add the omnibus_updater default cookbook to runlist on a windows 2012 r2 system with chef-client 12.6.0 already installed. The cookbook will run and try to update chef-client to the latest(12.7.2) version.

I tested clients 11.14.6 and 12.5.1 and everything worked just fine. Only 12.6.0 client throws access denied errors on a bunch of files during the uninstall phase. Then subsequent install phase errors out leaving a system without a chef-client installed. While logged on to the server and manually running chef-client from within powershell opened as administrator, I see a command line pop up window running the uninstall and several files saying access denied scroll by in the output.

I've tried adding work arounds by executing commands to change ownership, permissions, etc... and to delete the left over files but still get access denied on the files when running the commands from within the cookbook. Manual run of the commands in the command line prompt does work.

What's weird is while logged in and manually running the gui install and or removing files in explorer works just fine.
## Client Output

chef-client run
https://gist.github.com/Stromweld/dcc3a3b4a06a9048d39d

chef-client debug run
https://gist.github.com/Stromweld/bf80b9b8e427d5fb94de

msiexec log during install
https://gist.github.com/Stromweld/0aa0bee2d5d3461744d2
## Stacktrace

chef-stacktrace.out
https://gist.github.com/Stromweld/2f073ad11906dc85838d

**Comment by Stromweld** [comment]
experiencing the same behavior with machinges having 12.7.2 installed and trying to upgrade to 12.8.1. Manually running the MSI while RDP'ed to a server and you see a command line window pop up showing serveral files getting access denied during the removal phase. Using the gui installer though ignores the errors and continues with the install afterwards. However doing a quiet install from command line stops after the errors and doesn't start the install phase.

**Comment by thommay** [comment]
@chef/client-windows

**Comment by adamedx** [comment]
So if chef-client is calling a synchronous method to uninstall itself while still running....  that is generally not a good idea on Windows, and it may work, and it may not. If you look at agents like Windows Update or Intune, they trigger a separate process to do this. I'd suggest a different implementation than the current one -- rather than call the uninstall method inside of the chef-client run, just create a scheduled task that runs a script that does the same thing. That task can be triggered by the run and have some logic in it to shutdown the chef-client service if it's running and wait for any instances of chef to stop running for some timeout. There are various safeguards we could add to make it robust in the case of failure (i.e. it would be bad to end in a state where there is no chef-client on the box).

**Comment by Stromweld** [comment]
The scheduled task way just seems like a dirty work around and without being in the chef-client run you don't know if it worked or not etc... The main issue stems from the uninstall process throwing errors about access denied when trying to remove files in the opscode directory. Simmilar issue was raised back in October 2014. https://github.com/chef-cookbooks/windows/issues/130 

If you manually run the msi package while logged on to the windows server you also see a cmd line pop up window throw those errors as well during the uninstall phase of upgrading the chef-client software but will ignore them and continue with the install after the uninstall phase is done. But when trying the install directly from chef-client run or uninstall and then install as 2 seperate calls it errors out due to the files remaining. I've even tried to do uninstall, reset file permissions and calling to delete the files and folders then calling the install and it still fails due to the files not removing.

**Comment by Stromweld** [comment]
What changed in the MSI package between 12.5.1 and 12.6.0?

**Comment by emachnic** [comment]
So I'm not sure if this helps but I just ran into the "Access denied" errors  and after stopping the chef-client service, I was able to upgrade using the MSI GUI.

**Comment by Stromweld** [comment]
MSI gui always works, for some reason the failure happens when done from ruby during a chef-client run. The omnibus-updater cookbook I submitted the windows update portion and issue command to kill the service, then issue command to execute uninstall(prevents current chef-client run process from recieving kill signal), then issue msiexec command to install updated version. This last step is where it fails due to locked files. The process I wrote worked great all the way up to chef-client 12.6.0 when they started changing the msi package in preparation for seperate 32bit and 64bit clients on windows.

**Comment by AntoineBro** [comment]
seeing similar behavior updating from chef-client 12.7.2 to 12.12.15 on windows server 2008 r2

**Comment by jaymzh** [comment]
@adamedx So I understand what you're saying, which roughly amounts to "This is the usual model in Windows, it may be broken, but it's how it works" - and for sure we'll just have certain limitations on Windows. But let's not be too defeatist here. I have a few thoughts:

* regarding the argument of "but you can't upgrade software that's running" vis-a-vis stability. It's been a (very) long time since I've dug into the Windows linker, but it was pretty impressive last time I looked at it (had some features Linux lacked!), so as long as we follow the re-exec-on-upgrade pattern, there's no reason that ruby interpreter can't last the few milliseconds until Chef re-execs itself. To be clear the pattern on every other platform is as follows:

```
ruby_block 'restart chef' do
  block { exec('/usr/bin/chef-client') }
  action :nothing
end

package 'chef' do
  action :upgrade
  notifies :run, 'ruby_block[restart chef]', :immediately
end
```

I'm certain Windows can do this (minus default MSI stupidity) and we should strive to support it.

* External things - SCCM/WSUS and/or GroupPolicy and/or other things - can be used to upgrade Chef, but let's be clear, if you have to have some other thing to manage Chef that means Chef is failing at managing your machine. Even if this isn't our fault, we should work hard to make it not fail at this. Clearly this worked prior to 12.6 so we should see what changed and how can leverage what we were doing in 12.5 to make this work.

* As far as I can tell, scheduling tasks to upgrade oneself is a hack. I'm aware that SCCM/WSUS do this internally even to upgrade themselves, but it's a hack. It means a bunch of things: (1) we can't do it synchronously (2) Chef can't respond to the success/failure and (3) as you pointed out, great care would have to be taken to ensure we don't end up with no Chef.

* The live-action upgrade works, just not a silent-one... which ... seems to again imply something weird going on.

* Even in the event we want to just say we don't support this in any way ever (which I think is a cop-out, but I'm not a windows maintainer, so it's not up to me), this is still a regression. We should at least attempt to fix this and then formerly drop support in 13.

I feel like we can make this more delightful... but I understand I'm out an outsider on the Windows platform, so hopefully those are actually useful thoughts.

**Comment by svmastersamurai** [comment]
To throw in my two cents, I've been working on this as well. Needless to say this solution is... growing in complexity as opposed to being able to upgrade in-place 😆 

I have taken the suggestion of using a scheduled task that fires off an event to the event log that tells chef to upgrade itself, and the task is running as `NT AUTHORITY\SYSTEM`. Even still there are instances where a file lock is in place and the update of chef-client will fail. Once the chef-client fails to install we have the problem of chef-client _not being able to run at all_ because the local installation is completely gone except for the files that remain locked. Fun times! I've even started adding more scheduled tasks that verify chef-client can launch and then backing up the MSIs somewhere just so they can be reinstalled if the worst should happen. 

I have run `handle.exe` to try and see which processes have the lock on the file but for the most part that is coming up empty. If you wait 20-40 min the lock(s) are eventually released and the directory can be cleaned out. The file that I've seen locked most frequently in my testing has been the `chef-log.dll` which is part of the `win32-eventlog` gem.

A hacky fix is to just rename `C:\opscode\chef` to something else, this frees up the directory for the installer to extract everything to and the lock will go away on the renamed directory at some point in the future for you to cleanup the remnants. Automating this portion has been a pain though and I'm still constantly getting my development VM into the state where the orchestration has chef-client in an unrunnable state after the upgrade.

To echo @jaymzh it would be ideal to be able to mirror other OSes of doing an in-place upgrade if at all possible :)

**Comment by svmastersamurai** [comment]
Just to expand on some stuff I yammered about in previous comment I did a few things to attempt to figure out what is going on.

Here is a simple chef recipe I made to upgrade the chef-client after the run. Not very eloquent, but it works to test this process out.
https://gist.github.com/svmastersamurai/1d49eaad75195d33e29c864b7fd06ab7

I was curious what previous versions were like that made this upgrade process behave more smoothly. I went ahead and installed a few version and put a breakpoint into a recipe to see what modules were loaded.
https://gist.github.com/svmastersamurai/e41b4e2f303a47a658cd0a10e223d79f
As chef-client has matured on Windows leveraging libraries that allow ruby to interface more directly with the Windows API (e.g. FFI) is now the name of the game. Not necessarily a bad thing, mind you, but it comes with the cost of these files being locked up as the chef-client process runs.

## Analysis

Upgrades from chef-client 12.5.1 -> 12.6.0 were okay
chef-client 12.6.0 -> 12.7.2 I was able to trigger the access denied error. After firing up Procmon I captured the only remaining lock that was left when the upgrader ran. This aligns exactly with what was reported by @Stromweld
https://gist.github.com/svmastersamurai/a6df8cb8eedb4234850788914f384136

The lock on `chef-log.dll` remained there _for a few hours_ 😨 . In our environment we run chef maybe twice an hour, so this process would fail and leave the nodes unable to re-run chef.
The odd thing is that the DHCP Client seems to be the process holding the lock on the file.
https://gist.github.com/svmastersamurai/58ca5246a9cbf439591d6c5ac6f50e66

Dissecting the installer the cleanup for when the chef-client upgrades is just a call to `rd /s /q C:\opscode\chef`, not very much in the way of error handling here.

## Discussion

* Could the MSI do more?
  * Maybe pass a parameter to the MSI that will aggressively kills chef-client and then re-runs as part of the process?
  * Rename the old chef-client's folder so `C:\opscode\chef` is free for the installer to unzip files to.
    * We could clean-up remenants with chef but it would be nice for the installer to do it for us 😋 
    * Even if we do this though and there are locked files present that either chef-client or the installer could still error out on.
* What is the reason for the `chef-log.dll` being locked by the DHCP Client?
  * Why does the lock remain for so long after the chef run is completed?

**Comment by Stromweld** [comment]
Has anybody looked at this with the latest chef-client cookbook's switch from chef running as a service to now being a scheduled task on windows? I haven't had the time as work has me busy on other projects, but I'm hoping to be able to get back to experimenting on this again soon. Would be nice to have a seamless windows upgrade process during a chef-client run like on linux. That way if there is an error it gets reported back to chef server and is more noticeable.

**Comment by johnnyramos** [comment]
This issue seems more than just minor. The affects the possibility of upgrading the chef client in which can cause loss of a node.

**Comment by thommay** [comment]
@johnnyramos "Accepted Minor" means that we can fix the problem in the 12.x series (ie, with a minor version bump). It's not a statement on the severity of the issue.

**Comment by madhu2421** [comment]
It was very painful upgrading chef client versions to 12.x due to this issue. We weren't able to keep our infrastructure updated with latest versions of chef-client due to the issues while upgrading Chef client

**Comment by btm** [comment]
The simplest fix to unblock everyone, which involves renaming the chef directory before we upgrade: https://github.com/chef-cookbooks/omnibus_updater/pull/139/files

I've tested upgrading from both 12.5.1 and 12.15.19 to Chef 12.17.44 successfully.

@svanharmelen said:
> A hacky fix is to just rename C:\opscode\chef to something else, this frees up the directory for the installer to extract everything to and the lock will go away on the renamed directory at some point in the future for you to cleanup the remnants.

That's what this change does, but it's very simple and I don't consider it a hack. It is essential that we keep the existing files around in some way until the chef-client runs has complete. Windows will not allow us to delete them as long as the process is still running.

@svanharmelen said:
> I was curious what previous versions were like that made this upgrade process behave more smoothly.

By default, the Windows installer creates a system restore point when it installs software. This triggers the Volume Shadow Copy service to create a shadow copy. I'm guessing at this point that the Windows installer can then delete the files because the running processes file handles are point to the files in the shadow copy. However, I've tried running `chef-shell` and deleting the contents of the directory which then fails with access denied errors, then running `vssadmin create shadow /for=C:` and trying again and I still get the same access denied errors. So there's something else happening as well.

The FastMSI change put the contents of `c:\opscode\chef` in a zip file, which the MSI contains, and then we unzip it as a WIX CustomAction as part of the installer. Previously the MSI contained all of the files, and the Windows Installer would manage each individual file in such a way as to provide other features like rollbacks on failed upgrades, repairs, major vs minor upgrades. On the surface this is great, but because we ship so many files inside the Omnibus packages it lead to exorbitant install times, especially on upgrades where uninstalling takes just as long as installing.

We can continue investigating trying to duplicate the shadow copy functionality or whatever the windows installer is using, but I suggest using these changes to omnibus_updater for now.

**Comment by btm** [comment]
https://msdn.microsoft.com/en-us/library/windows/desktop/aa372060(v=vs.85).aspx

> The installer only creates checkpoints in these cases when the change is run with at least a basic UI.

If no restore point is being created on unattended installs, the shadow copy theory is out.

Running an install by hand with `msiexec /i PATH_TO_MSI /qn /log install.log` and then an upgrade with a similar `msiexec /i PATH_TO_NEWER_MSI /qn /log upgrade.log` we see a line in the second `upgrade.log` file like:

```
Info 1903. Scheduling reboot operation: Deleting file D:\Config.Msi\2841496c.rbf. Must reboot to complete operation.
```

https://blogs.msdn.microsoft.com/windows_installer_team/2005/09/18/insights-into-the-underlying-windows-installer-reboot-behavior/

This blog post indicates that what is happening is because we're doing a silent install we're ignoring the fact that the file is in use and the Windows Installer has automatically configured our system to replace the in use file with the newer file **on the next reboot**.

What appeared prior to 12.6 to be providing a seamless in-place upgrade was actually one that somewhat silently required a reboot afterwards to put the system in the expected state. If you're regularly rebooting anyway that's probably okay, but I doubt that requirement is what most people want.

Unless I'm missing some magic with the directory move in-place upgrade suggested in https://github.com/chef-cookbooks/omnibus_updater/pull/139, that looks like a superior pattern than what was happening prior to 12.6 because it doesn't require a reboot.

For everyone else who isn't attach to in-place upgrade in the middle of a chef-client run, leveraging a scheduled task managed by a chef-client run that later stops the service and performs the upgrade is likely the best way to ensure your upgrade on disk is complete.
